src/client/executor.rs

```diff
@@ -583,8 +584,11 @@ impl Client {
         // sessions are supported or not.
         match initial_status {
             SessionSupportStatus::Undetermined => {
+                let topology_type = self.inner.topology.topology_type().await;
                 let criteria = SelectionCriteria::Predicate(Arc::new(move |server_info| {
-                    server_info.server_type().is_data_bearing()
+                    let server_type = server_info.server_type();
+                    (matches!(topology_type, TopologyType::Single) && server_type.is_available())
```

Comment:
for direct connections and standalones, just try to select any server. for all other topologies, wait for a data bearing server.

Reply:
We use this same logic in `transaction_support_status` directly below, so if a user is trying to execute a transaction on a direct connection to a hidden node I believe they'll encounter the same problem (although this is a much more niche use case). Can we make these changes for `TransactionSupportStatus` as well?

Reply:
Good catch, fixed.